### PR TITLE
Make data challenge harder

### PR DESCRIPTION
**Pull request by SashaWeinstein**
# Harder data challenge 😈 
I realized once I was halfway done that we may want to move this challenge to a new repo instead of changing the existing readme, figured the changelog in the PR might still be a good way to view the changes. 

My thoughts process on this change:
- I liked [this data challenge Andrew shared](https://github.com/mysidewalk/interview/blob/master/backend-engineer/README.md) that tells you that all the code needs to be runnable from a single bash script. Passing command line args is a pretty big part of what we do so it makes sense to have the interviewee do that
- I had them choose their own administrative boundaries to test their googling skills and knowledge of filetypes. If they pull from some weird file that's not a geojson, csv or shapefile that's not a good sign
- I didn't tell them what to name their files or how many to save as good file naming is important and I want to see if they can explain their choices well

Looking forward to discussing more!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
+### Python Task 2: Data Aggregation
 
-Create a time series table based on the `data/raw.csv` file we created from **Task 1** that has the following fields
+Write a process to produce a time series table based on the data you've pulled so far. It should have the following fields
 
-- `created_date_hour`: the timestap of request creation by date and hour
-- `complaint_type`: the type of the complaint
+- `created_time`: the timestap of request creation by date and hour OR just date
+- `complaint_type`: the type of the complaint (optional)
 - `count`: the count of service requests by `complaint_type` by `created_date_hour`
```
Comment (AmandaDoyle) [inline README.md:63]: If `complaint_type` is optional then this should be recast.  Also, `created_date_hour` needs to be changed to `created_time`

File: README.md [hunk 2]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
 
-Store this table in a csv under the `data` folder with a csv file name of your choice.
+The granularity of the `created_time` column and whether the data is broken down by complaint type should be controlled by parameters passed in from the command line.
```
Comment (AmandaDoyle) [inline README.md:65]: This needs to be moved up top, to explicitly state the you should be able to pass parameters through the process.

File: README.md [hunk 3]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
+If you would like to go back and download different tables to visualize, that's totally fine. Chose whichever pulls you think make it easiest to show us your work. 
 
-At Data Engineering, we enhance datasets with geospatial attributes, such as point locations and administrative boundaries. To help us better understand the data from **Task 1**, we would like you to join the initial raw data to the **[2020 NTA (Neighborhood Tabulation Area) boundaries](https://www1.nyc.gov/site/planning/data-maps/open-data/census-download-metadata.page)** and create a choropleth map of 7 day total count by NTA of a specific `complaint_type` of your choice.
+### Python Task 4: Spatial data processing
 
-Depending on how you generate the map, you can store the map as a `.png` or `.html` under the `data` folder.
+At Data Engineering, we enhance datasets with geospatial attributes, such as point locations and administrative boundaries. To help us better understand the data from **Python Task 1**, we would like you to join the initial raw data to the some NYC administraive boundary of your chosing. Then create a choropleth map of count of complaints of a specific `complaint_type` of your choice.
```
Comment (AmandaDoyle) [inline README.md:76]: change `some NYC administraive boundary` to `an NYC administraive boundary`
Given how prevalent `complaint_type` is throughout the challenge, I don't think that we can make it optional (i.e. complaints of a specific `complaint_type` of your choice.)

File: README.md [hunk 4]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
 
-- Load the `data/raw.csv` into a database of your choice and name the table `sample_311`. Make sure this process is captured in a script.
-- Perform the same aggregation in **Task 2** in SQL and store the results in a table (same name as the corresponding csv file).
+### SQL/Docker Task 1: Build container and load 
+
+We ❤️ SQL and docker! At Data Engineering, we work with database containers a lot and we write a lot of fast and simple ETL pipelines using SQL. In this task, you will:
 
-> Note: Depending on your preference, you can use or [Postgres](https://www.postgresql.org/), which is prefered; however, if you are familiar with [SQLite](https://docs.python.org/3/library/sqlite3.html) (much easier to set up and use), you can use that too.
+- Set up POSTGIS container using an image
+- Load the `data/raw.csv` into the database and name the table `sample_311`. Make sure this process is captured in a script.
```
Comment (AmandaDoyle) [inline README.md:86]: You didn't tell them how to name files up top, so `data/raw.csv` is out of place here.  Naming is hard.  I'm okay with providing some guidance to make things easier to explain downstream.

File: README.md [hunk 5]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
+- Perform the same aggregation in **Task 2** in SQL and store the results in a table (same name as the corresponding csv file).
 
-### Task 6: Spatial SQL
+### SQL/Docker Task 2: Spatial SQL
 
 A lot of popular databases have geospatial extensions, which makes spatial data processing in SQL super easy to use. In this task you will:
 
-- Load the NTA data to the database as a spatial table
-- Do a spatial join in SQL between `sample_311` and the NTA table and add a `nta` column to `sample_311`
+- Load the adminstrative data you used in python task 4 to the database as a spatial table
```
Comment (AmandaDoyle) [inline README.md:93]: `python task 4` should be in bold like other referenced tasks

File: README.md [hunk 6]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
+- Do a spatial join in SQL between `sample_311` and the administrative boundaries and add a corresponding column to `sample_311`
```
Comment (AmandaDoyle) [inline README.md:94]: Is the corresponding the unique ID of the admin boundary?  If so, I'd explicitly say that.

File: README.md [hunk 7]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
-We ❤️ SQL! At Data Engineering, we deal with databases a lot and we write a lot of fast and simple ETL pipelines using SQL. In this task, you will:
 
-- Load the `data/raw.csv` into a database of your choice and name the table `sample_311`. Make sure this process is captured in a script.
-- Perform the same aggregation in **Task 2** in SQL and store the results in a table (same name as the corresponding csv file).
+### SQL/Docker Task 1: Build container and load 
+
+We ❤️ SQL and docker! At Data Engineering, we work with database containers a lot and we write a lot of fast and simple ETL pipelines using SQL. In this task, you will:
 
-> Note: Depending on your preference, you can use or [Postgres](https://www.postgresql.org/), which is prefered; however, if you are familiar with [SQLite](https://docs.python.org/3/library/sqlite3.html) (much easier to set up and use), you can use that too.
+- Set up POSTGIS container using an image
```
Comment (AmandaDoyle) [inline README.md:85]: Should we point to one of our docker images?  Or point them to a docker image?

**Review by AmandaDoyle — CHANGES_REQUESTED** [review]
@SashaWeinstein  see my comments, which are mostly around wording to provide more clarity.  Happy to talk through anything.  I think this can do on a new branched named `advanced` or `senior` or?

File: README.md [hunk 8]
```diff
@@ -42,51 +42,59 @@ Your code interview will be evaluated based on your repo, so make sure all files
...
+The logic to perform each step should be written in python, and be callable by an executable script that takes the arguments.  
+
 We love the NYC 311 service and the open data products that come with it. In this challenge, you will use **[311 Service Requests from 2010 to Present](https://data.cityofnewyork.us/Social-Services/311-Service-Requests-from-2010-to-Present/erm2-nwe9)** on NYC Open Data, write an ETL pipeline, and produce some data insight.
 
-### Task 1: Data Download
 
-Write a python script/notebook to download all service request records created in the **last week** (7 days) and has **HPD** as the responding agency, and store the data in a csv named `raw.csv` in a folder called `data`.
+### Python Task 1: Data Download
+
+Write a process to download all service request records created in the last *x* days with a certain responding agency. It should take the number of days and responding agency acronym as parameters passed from the command line. Save the service requests in the last seven days with HPD as the responding agnecy and a second different pull of the data in two adequatly named csv files in a folder called `data`.
```
[the remark this answers is not in the document]
Comment (SashaWeinstein) [reply README.md:55]: Yes, to show us that passing the responding agency as argument works

File: README.md (hunk 1 above)
Comment (SashaWeinstein) [reply README.md:63]: The `created_date_hour` column was renamed. I included "(optional)" after the description of complaint type, is there a better way to communicate that?

File: README.md (hunk 3 above)
Comment (SashaWeinstein) [reply README.md:76]: Oh I meant optional w/r/t the code, not optional within the scope of the data challenge. I think it's a common pattern to leave an argument out and have the code default to "all". We can be clearer that the interviewee needs to save at least one file that does break down the data by complaint type

File: README.md (hunk 4 above)
Comment (SashaWeinstein) [reply README.md:86]: Oh yea I can change that

File: README.md (hunk 6 above)
Comment (SashaWeinstein) [reply README.md:94]: Good question. I think we could say "administrative boundary ID" which is easier or "foreign key" to test their knowledge of SQL terms/googling ability

File: README.md (hunk 7 above)
Comment (SashaWeinstein) [reply README.md:85]: I think we should be more specific, like say "official open-source postgis image" but make them google to find [the image we use](https://registry.hub.docker.com/r/postgis/postgis/)

**Pull request closed without merging**